src/lib/textToEdit.ts

```diff
@@ -0,0 +1,351 @@
+import { Models } from '@kittycad/lib'
+import { VITE_KC_API_BASE_URL } from 'env'
+import crossPlatformFetch from './crossPlatformFetch'
+import { err, reportRejection } from './trap'
+import { Selections } from './selections'
+import { ArtifactGraph, getArtifactOfTypes } from 'lang/std/artifactGraph'
+import { SourceRange } from 'lang/wasm'
+import toast from 'react-hot-toast'
+import { codeManager, editorManager, kclManager } from './singletons'
+import { ToastPromptToEditCadSuccess } from 'components/ToastTextToCad'
+import { uuidv4 } from './utils'
+import { diffLines } from 'diff'
+import { Transaction, EditorSelection, SelectionRange } from '@codemirror/state'
+import { modelingMachineEvent } from 'editor/manager'
+
+function sourceIndexToLineColumn(
+  code: string,
+  index: number
+): { line: number; column: number } {
+  const codeStart = code.slice(0, index)
+  const lines = codeStart.split('\n')
+  const line = lines.length
+  const column = lines[lines.length - 1].length
+  return { line, column }
+}
+
+function convertAppRangeToApiRange(
+  range: SourceRange,
+  code: string
+): Models['SourceRange_type'] {
+  return {
+    start: sourceIndexToLineColumn(code, range[0]),
+    end: sourceIndexToLineColumn(code, range[1]),
+  }
+}
+
+export async function submitPromptToEditToQueue({
+  prompt,
+  selections,
+  code,
+  token,
+  artifactGraph,
+}: {
+  prompt: string
+  selections: Selections
+  code: string
+  token?: string
+  artifactGraph: ArtifactGraph
+}): Promise<Models['TextToCadIteration_type'] | Error> {
+  const ranges: Models['TextToCadIterationBody_type']['source_ranges'] =
+    selections.graphSelections.flatMap((selection) => {
+      const artifact = selection.artifact
+      const prompts: Models['TextToCadIterationBody_type']['source_ranges'] = []
+
+      if (artifact?.type === 'cap') {
+        prompts.push({
+          prompt: `The users main selection is the end cap of a general-sweep (that is an extrusion, revolve, sweep or loft).
+The source range most likely refers to "startProfileAt" simply because this is the start of the profile that was swept.
+If you need to operate on this cap, for example for sketching on the face, you can use the special string ${
+            artifact.subType === 'end' ? 'END' : 'START'
+          } i.e. \`startSketchOn(someSweepVariable, ${
+            artifact.subType === 'end' ? 'END' : 'START'
+          })\`
+When they made this selection they main have intended this surface directly or meant something more broad like the sweep itself.
```

Comment:
```suggestion
When they made this selection they main have intended this surface directly or meant something more general like the sweep body.
```
brb adding "prompt engineer" to my LinkedIn

Reply:
Added this.